sound/soc/soc-pcm.c

```diff
@@ -2015,23 +2016,33 @@ int dpcm_be_dai_trigger(struct snd_soc_pcm_runtime *fe, int stream,
 		dev_dbg(be->dev, "ASoC: trigger BE %s cmd %d\n",
 			be->dai_link->name, cmd);
 
+		do_trigger = false;
 		switch (cmd) {
 		case SNDRV_PCM_TRIGGER_START:
 			spin_lock_irqsave(&fe->card->dpcm_lock, flags);
-			if ((be->dpcm[stream].state != SND_SOC_DPCM_STATE_PREPARE) &&
+			if (!be->dpcm[stream].be_start &&
+			    (be->dpcm[stream].state != SND_SOC_DPCM_STATE_PREPARE) &&
 			    (be->dpcm[stream].state != SND_SOC_DPCM_STATE_STOP) &&
 			    (be->dpcm[stream].state != SND_SOC_DPCM_STATE_PAUSED)) {
 				spin_unlock_irqrestore(&fe->card->dpcm_lock, flags);
 				continue;
 			}
 			state = be->dpcm[stream].state;
+			be->dpcm[stream].be_start++;
+			if (be->dpcm[stream].be_start == 1)
+				do_trigger = true;
+
 			be->dpcm[stream].state = SND_SOC_DPCM_STATE_START;
 			spin_unlock_irqrestore(&fe->card->dpcm_lock, flags);
 
+			if (!do_trigger)
+				continue;
```

Comment:
Will it be better to keep be->dpcm[stream].state unchanged when `soc_pcm_trigger` is not called?

Reply:
No, this is really to keep track of the number of requests to start/stop. We change the state to START when the first request is made, and conversely we change the state to STOP when the last request is made.

Reply:
@plbossart `do_trigger` will be true at first request and we will set `be->dpcm[stream].state = SND_SOC_DPCM_STATE_START` then call `soc_pcm_trigger`. `do_trigger` will be false and we will not call `soc_pcm_trigger` at second request. Ideally, `be->dpcm[stream].state` should be already START and we don't need to set it to START again at the second request. My point is that will it cause any problem if do_trigger = false and `be->dpcm[stream].state` is not START. As a result, `be->dpcm[stream].state` will be set as START but `soc_pcm_trigger` is actually not called in this corner case.

Reply:
Sorry @bardliao I don't see how we can have a state with 'do_trigger==false' and 'be->dpcm[stream].state != START'.
Can you elaborate on the scenario?

Reply:
@plbossart I can't image a scenario, but the existing code changes be->dpcm[stream].state after soc_pcm_trigger().

Good news is that I found an issue while I was trying to find a scenario that `do_trigger==false` and `be->dpcm[stream].state != START`. I will see IO error if I do the following test.
1. aplay -Dhw:0,0 wav/LR_48k.wav -vv -i
2. Pause hw:0,0
3. Create a new terminal and run aplay -Dhw:0,31 wav/LR_48k.wav -vv -i
4. Stop hw:0,31
5. Release hw:0,0 ==> IO error

See [dmesg_new.txt](https://github.com/thesofproject/linux/files/7038953/dmesg_new.txt) with below change

```patch
diff --git a/sound/soc/soc-pcm.c b/sound/soc/soc-pcm.c
index 1ece4bd32143..53310a06a89a 100644
--- a/sound/soc/soc-pcm.c
+++ b/sound/soc/soc-pcm.c
@@ -1018,6 +1018,7 @@ static int soc_pcm_trigger(struct snd_pcm_substream *substream, int cmd)
        int ret = -EINVAL, _ret = 0;
        int rollback = 0;

+       dev_info(rtd->dev, "bard: %s cmd %d\n", __func__, cmd);
        switch (cmd) {
        case SNDRV_PCM_TRIGGER_START:
        case SNDRV_PCM_TRIGGER_RESUME:
@@ -2024,8 +2025,8 @@ int dpcm_be_dai_trigger(struct snd_soc_pcm_runtime *fe, int stream,
                if (!snd_soc_dpcm_be_can_update(fe, be, stream))
                        continue;

-               dev_dbg(be->dev, "ASoC: trigger BE %s cmd %d\n",
-                       be->dai_link->name, cmd);
+               dev_dbg(be->dev, "ASoC: trigger BE %s cmd %d be_start %d be->dpcm[stream].state %d\n",
+                       be->dai_link->name, cmd, be->dpcm[stream].be_start, be->dpcm[stream].state);

                do_trigger = false;
                switch (cmd) {
@@ -2043,6 +2044,8 @@ int dpcm_be_dai_trigger(struct snd_soc_pcm_runtime *fe, int stream,
                        if (be->dpcm[stream].be_start == 1)
                                do_trigger = true;

+                       dev_dbg(be->dev, "bard: ASoC: trigger BE %s cmd %d be_start %d do_trigger %d\n",
+                               be->dai_link->name, cmd, be->dpcm[stream].be_start, do_trigger);
                        be->dpcm[stream].state = SND_SOC_DPCM_STATE_START;
                        spin_unlock_irqrestore(&fe->card->dpcm_lock, flags);

```
And there is no issue without this PR
See [dmesg_old.txt](https://github.com/thesofproject/linux/files/7038959/dmesg_old.txt) with
```patch
diff --git a/sound/soc/soc-pcm.c b/sound/soc/soc-pcm.c
index fc1854e3e43f..b834e510646d 100644
--- a/sound/soc/soc-pcm.c
+++ b/sound/soc/soc-pcm.c
@@ -1018,6 +1018,7 @@ static int soc_pcm_trigger(struct snd_pcm_substream *substream, int cmd)
        int ret = -EINVAL, _ret = 0;
        int rollback = 0;

+       dev_err(rtd->dev, "bard: %s cmd %d\n", __func__, cmd);
        switch (cmd) {
        case SNDRV_PCM_TRIGGER_START:
        case SNDRV_PCM_TRIGGER_RESUME:
@@ -2010,8 +2011,8 @@ int dpcm_be_dai_trigger(struct snd_soc_pcm_runtime *fe, int stream,
                if (!snd_soc_dpcm_be_can_update(fe, be, stream))
                        continue;

-               dev_dbg(be->dev, "ASoC: trigger BE %s cmd %d\n",
-                       be->dai_link->name, cmd);
+               dev_dbg(be->dev, "ASoC: trigger BE %s cmd %d be->dpcm[stream].state %d\n",
+                       be->dai_link->name, cmd, be->dpcm[stream].state);

                switch (cmd) {
                case SNDRV_PCM_TRIGGER_START:
```

The behavior deference is that TRIGGER_STOP will not be triggered in step 4 in the existing code. `snd_soc_dpcm_can_be_free_stop` do an important job here.

Reply:
Valid point @bardliao , I added a patch to deal with the TRIGGER_PAUSE_RELEASE. This should also work when the BE is in STOP and START states, depending on what the FEs do.
I think the logic only worked when there was no mixer...